apis/v1alpha2/clusternetworkpolicy_types.go

```diff
@@ -0,0 +1,479 @@
+	// access, DomainNames Allow rules should be used with a lower priority
+	// egress deny -- this allows the admin to maintain an explicit "allowlist"
+	// of reachable domains.
+	//
+	// DomainNames can have up to 25 domain names specified in one rule.
+	//
+	// <network-policy-api:experimental>
+	// +optional
+	// +listType=set
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=25
+	DomainNames []DomainName `json:"domainNames,omitempty"`
+}
+
+// DomainName describes one or more domain names to be used as a peer.
+//
+// DomainName can be an exact match, or use the wildcard specifier '*' to match
+// one or more labels.
+//
+// '*', the wildcard specifier, matches one or more entire labels. It does not
+// support partial matches. '*' may only be specified as a prefix.
+//
+// Examples:
+//   - `kubernetes.io` matches only `kubernetes.io`.
+//     It does not match "www.kubernetes.io", "blog.kubernetes.io",
+//     "my-kubernetes.io", or "wikipedia.org".
+//   - `blog.kubernetes.io` matches only "blog.kubernetes.io".
+//     It does not match "www.kubernetes.io" or "kubernetes.io".
+//   - `*.kubernetes.io` matches subdomains of kubernetes.io.
+//     "www.kubernetes.io", "blog.kubernetes.io", and
+//     "latest.blog.kubernetes.io" match, however "kubernetes.io", and
+//     "wikipedia.org" do not.
+//
+// +kubebuilder:validation:Pattern=`^(\*\.)?([a-zA-z0-9]([-a-zA-Z0-9_]*[a-zA-Z0-9])?\.)+[a-zA-z0-9]([-a-zA-Z0-9_]*[a-zA-Z0-9])?\.?$`
+type DomainName string
+
+// +kubebuilder:object:root=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterNetworkPolicyList contains a list of ClusterNetworkPolicy
+type ClusterNetworkPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ClusterNetworkPolicy `json:"items"`
+}
+
+// ClusterNetworkPolicySubject defines what resources the policy applies to.
+// Exactly one field must be set.
+// +kubebuilder:validation:MaxProperties=1
+// +kubebuilder:validation:MinProperties=1
+type ClusterNetworkPolicySubject struct {
+	// Namespaces is used to select pods via namespace selectors.
+	// +optional
+	Namespaces *metav1.LabelSelector `json:"namespaces,omitempty"`
+	// Pods is used to select pods via namespace AND pod selectors.
+	// +optional
+	Pods *NamespacedPod `json:"pods,omitempty"`
+}
+
+// NamespacedPod allows the user to select a given set of pod(s) in
+// selected namespace(s).
+type NamespacedPod struct {
+	// NamespaceSelector follows standard label selector semantics; if empty,
+	// it selects all Namespaces.
+	// it selects all Namespaces.
+	NamespaceSelector metav1.LabelSelector `json:"namespaceSelector"`
+
+	// PodSelector is used to explicitly select pods within a namespace;
+	// if empty, it selects all Pods.
+	PodSelector metav1.LabelSelector `json:"podSelector"`
+}
+
+// ClusterNetworkPolicyPort describes how to select network ports on pod(s).
```

Comment:
"how to select destination network ports".

for egress Nodes and Networks rules this describes ports on hosts, not on pods.

(The ANP docs apparently never got updated to describe this correctly.)